### PR TITLE
Add aad b2c filter, #464

### PR DESCRIPTION
**Pull request by Incarnation-p-lee**

### CODE REVIEW

File: azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/b2c/AADB2CFilter.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+    public AADB2CFilter(@NonNull AADB2CProperties b2cProperties,
+                        @NonNull AADB2CLogoutSuccessHandler logoutSuccessHandler) {
+        super();
+
+        this.logoutSuccessHandler = logoutSuccessHandler;
+        this.signUpOrInRedirectURL = b2cProperties.getPolicies().getSignUpOrSignIn().getRedirectURI();
+    }
+
+    public String getLogoutSuccessURL() {
```
Comment (xscript) [inline azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/b2c/AADB2CFilter.java]: Just think of that whether we support relative path? Does Spring Security support relative path as successful logout URL?

File: azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/b2c/AADB2CFilterScenario.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    private AADB2CFilterScenarioHandler scenarioHandler;
+
+    AADB2CFilterScenario(@NonNull AADB2CFilterScenarioHandler scenarioHandler) {
+        this.scenarioHandler = scenarioHandler;
+    }
+
+    public static AADB2CFilterScenario resolve(@NonNull HttpServletRequest request, @NonNull AADB2CFilter filter) {
+        final String requestURL = request.getRequestURL().toString();
+
+        if ("GET".equals(request.getMethod()) && requestURL.equals(filter.getSignUpOrInRedirectURL())) {
```
Comment (xscript) [inline azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/b2c/AADB2CFilterScenario.java]: Is there built-in enum/constant for HTTP method names?

File: azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/b2c/AADB2CFilter.java (hunk 1 above)
Comment (Incarnation-p-lee) [reply azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/b2c/AADB2CFilter.java]: Yes, it support relative path, need to enhance this part for SuccessHandler.

File: azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/b2c/AADB2CFilterScenario.java (hunk 2 above)
Comment (Incarnation-p-lee) [reply azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/b2c/AADB2CFilterScenario.java]: yes, should leverage `org.springframework.http.HttpMethod`.

File: azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/b2c/AADB2CURL.java [hunk 3]
```diff
@@ -48,11 +46,23 @@ private static String getEncodedURL(@NonNull String url) {
...
         try {
             new URL(url);
-        } catch (MalformedURLException e) {
-            throw new AADB2CConfigurationException("Invalid URL: " + url, e);
+            return url;
+        } catch (MalformedURLException ignore) {
+            try {
+                new URI(url);
+                // Absolute URL example: http://localhost:8080/greeting
+                return String.format("%s://%s:%d/%s",
```
Comment (xscript) [inline azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/b2c/AADB2CURL.java]: As synced offline, let's make sure there is no gap to build URL by ourselves.

File: azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/b2c/AADB2CFilterScenario.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+    AADB2CFilterScenario(@NonNull AADB2CFilterScenarioHandler scenarioHandler) {
+        this.scenarioHandler = scenarioHandler;
+    }
+
+    public static AADB2CFilterScenario resolve(@NonNull HttpServletRequest request, @NonNull AADB2CFilter filter) {
+        final String requestURL = request.getRequestURL().toString();
+        final String method = request.getMethod();
+
+        if (HttpMethod.GET.matches(method) && requestURL.equals(filter.getSignUpOrInRedirectURL())) {
```
Comment (xscript) [inline azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/b2c/AADB2CFilterScenario.java]: Ignore case equals?

Comment (Incarnation-p-lee) [reply azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/b2c/AADB2CFilterScenario.java]: No need here, as Method names are case-sensitive, `get` is another method like user defined.

Comment (xscript) [reply azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/b2c/AADB2CFilterScenario.java]: What about `requestURL`?

Comment (Incarnation-p-lee) [reply azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/b2c/AADB2CFilterScenario.java]: In fact, only domain name is case insenstive, so we cannot do that here.
For example, 
https://en.wikipedia.org/wiki/Case_sensitivity
https://en.wikipedia.org/wiki/CASE_SENSITIVITY
Are totally different pages.

File: azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/b2c/AADB2CFilterLogoutSuccessHandler.java [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+@NoArgsConstructor
+public class AADB2CFilterLogoutSuccessHandler implements AADB2CFilterScenarioHandler {
+
+    @Override
+    public void handle(HttpServletRequest request, HttpServletResponse response) {
+        throw new UnsupportedOperationException("AADB2CFilterLogoutSuccessHandler implemented yet.");
```
Comment (sophiaso) [inline azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/b2c/AADB2CFilterLogoutSuccessHandler.java]: Error message incorrect? `implemented` -> `not implemented`?

Comment (Incarnation-p-lee) [reply azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/b2c/AADB2CFilterLogoutSuccessHandler.java]: yes, will remove that in next PR.

**Pull request merged**